### ISSUE TITLE
bug: RStudio Connection Panel Issues with ODBC drivers on M1/M2 macs

### ISSUE THREAD

**Issue by coatless** [rstudio/rstudio#12683]
### System details

    RStudio Edition : Desktop
    RStudio Version :  2022.12.0+353
    OS Version      :  macOS
    R Version       :  4.2.2

### Steps to reproduce the problem

1. Install ODBC drivers using steps at:
https://solutions.posit.co/connections/db/best-practices/drivers/#apple-macos 

2. Setup configuration files under local aliases: 

https://solutions.posit.co/connections/db/best-practices/drivers/#setting-up-database-connections-1 

Only modify the `[SQLite Driver]` location for `odbcinst.ini` file.

In place of driver location on Intel macs, the driver is:

```bash
/usr/local/lib/libsqlite3odbc.dylib
```

Thus, the location for the SQLite driver on M1/M2 macs is: 

```bash
/opt/homebrew/lib/libsqlite3odbc.dylib
```

3. In RStudio, manually install the `odbc` package:

```r
install.packages("odbc")
```

5. Try to create a new connection using the Connections Panel in the top-right

At this point, the Connections menu only shows `SQLite` not the `SQLIte Driver`. 

<img width="503" alt="m1-mac-only-shows-sqlite-option" src="https://user-images.githubusercontent.com/833642/217720005-861c2e4c-859a-440c-b9b9-1f9d76aca05b.png">

Select the `SQLite` option. 

Press Okay to create the connection

6. Receive the error of:

```
[unixODBC][Driver Manager]Data source name not found, and no default driver specified
```

<img width="592" alt="odbc-screenshot-error" src="https://user-images.githubusercontent.com/833642/217719786-5c967be4-b0c7-403b-a42a-b6f00862c173.png">

### Describe the problem in detail

The ODBC drivers offered by Homebrew for an M1/M2-powered mac seem to not be registered by RStudio. The Intel drivers appear to work as expected.

### Describe the behavior you expected

After the configuration files are setup, then RStudio should display the `SQLite Driver` option in the connections panel.

<img width="615" alt="rstudio-sqlite-driver-option-intel-mac" src="https://user-images.githubusercontent.com/833642/217719882-a086c720-701b-43f8-b328-5e12f72105ba.PNG">

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [x] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@coatless Thank you for raising this! 

@jonvanausdeln Can you take a look at this, since it requires an M1 or M2 Mac?

**Comment by jonvanausdeln** [comment]
I just tried with Elsbeth on my M1.. and yes, I can reproduce and see the exact same error.

**Comment by jonvanausdeln** [comment]
Note.. I can only get the driver to show up if I put the ini files in my home directory

`~/.odbc.ini`
`~/.odbcinst.ini`

As hidden files

**Comment by sharon-wang** [comment]
@jonvanausdeln @coatless if you have a chance, could you check that the Data Source Name being used is correct? I only hit the `Data source name not found and no default driver specified` error when using an incorrect DSN. More details below 👇 

---

Same as @jonvanausdeln, I filled out my `~/.odbc.ini` and `~/.odbcinst.ini`.

`~/.odbc.ini`
```
[SQLite]
Driver      = SQLite Driver
Database    = /tmp/testing
```
`~/.odbcinst.ini`
```
[SQLite Driver]
Driver          = /opt/homebrew/lib/libsqlite3odbc.dylib
```
As a result, I see both SQLite ad SQLite Driver in my New Connection list:

<img width="527" alt="image" src="https://user-images.githubusercontent.com/25834218/221030310-c21e37d3-c7f8-46c0-af70-17d1e9e3d7a3.png">

From this dialog, if I choose "SQLite" or "SQLite Driver" and click `Test` or `OK`, I get a different error:
```
Error: nanodbc/nanodbc.cpp:1118: 00000: [unixODBC][Driver Manager]Can't open lib '/opt/homebrew/lib/libsqlite3odbc.dylib' : file not found
```
- this error is bizarre because the dylib is in fact in that location. I even tried setting the Driver location in `~/.odbcinst.ini` to the actual homebrew Cellar location, but I still got the same error.

I can only reproduce the `nanodbc/nanodbc.cpp:1118: 00000: [unixODBC][Driver Manager]Data source name not found and no default driver specified ` error if I use an invalid Data Source Name.

### Demo of what I'm experiencing:

https://user-images.githubusercontent.com/25834218/221033595-96307961-c911-488c-9ace-ca99f6d0fc0e.mov

---

Some helpful commands I used to check my odbc configuation:
- Show Drivers
    - R Console: `odbc::odbcListDrivers()`
    - Terminal: `odbcinst -q -d`
- Show Data Source Names (DSNs)
    - Terminal: `odbcinst -q -s`
        - note that the source name printed from this command should be used in this dialog as the Data Source Name

**Comment by jonvanausdeln** [comment]
> @jonvanausdeln @coatless if you have a chance, could you check that the Data Source Name being used is correct? I only hit the `Data source name not found and no default driver specified` error when using an incorrect DSN. More details below 👇
> 

I just double checked and they data source names look correct

```
(test) Jons-MacBook-Pro:~ jonvanausdeln$ odbcinst -q -s
[PostgreSQL]
[SQLite]

```

And I get the error: 

```
> library(DBI)
> con <- dbConnect(odbc::odbc(), "SQLite", timeout = 10)
Error: nanodbc/nanodbc.cpp:1118: 00000: [unixODBC][Driver Manager]Data source name not found and no default driver specified 

```

**Comment by coatless** [comment]
I'm sorry, but I do not have an ARM-based mac. So, I can't test further. The initial debugging was with a student who had an M1 mac.

**Comment by ronblum** [comment]
This looks like the error message if the driver name doesn't match one in the `odbcinst.ini` file, although if I recall, @jonvanausdeln your environment has definitions for these in either the local or global file.

**Comment by sharon-wang** [comment]
@ronblum Jon and I compared our local environments and had the same versions for `R`, `unixodbc` and `sqliteodbc`, and we both have definitions in our driver config files. One difference is that Jon's mac is on MacOS 12.6 and mine is on MacOS 13.2; however, it's not yet apparent how/if that might relate to our difference in errors.

**Comment by sharon-wang** [comment]
Trevor also repro'd this issue and got the `Error: nanodbc/nanodbc.cpp:1118: 00000: [unixODBC][Driver Manager]Data source name not found and no default driver specified` message. Same `~/.odbc.ini` and `~/.odbcinst.ini` setup as me, and with the following versions:

```
sqliteodbc 0.9998
unixodbc 2.3.11
macOS Monterey (12.6)
```

**Comment by rantoun** [comment]
I was running into this issue as well and banging my head against the wall. I tried all the same steps and had the same issues as @coatless. The fix that ended up working for me was to set `ODBCSYSINI` env variable to the correct location of files (for me it is in `/opt/homebrew/etc`) in **`~/.Renviron`**. 

Once I did that and restarted R, I was able to see the drivers using `odbc::odbcListDrivers()` and then was able to connect.

**Comment by sharon-wang** [comment]
@rantoun thanks for sharing your steps!

I've set up a VM with MacOS 12.6 and can reproduce the error `Error: nanodbc/nanodbc.cpp:1118: 00000: [unixODBC][Driver Manager]Data source name not found and no default driver specified`.

Following @rantoun's steps, I was able to get the connection working for MacOS 12.6:

### M1 MacOS 12.6
1. ensure `usethis` package is installed in RStudio
2. R Console: `usethis::edit_r_environ()` to open `.Renviron` file in RStudio
3. Add `ODBCSYSINI=/opt/homebrew/etc` to the `.Renviron` file and save the file
4. Session > Restart R
5. ensure odbc config files are set up correctly
    ```
    $ cat ~/.odbc.ini
    [SQLite]
    Driver      = SQLite Driver
    Database    = /tmp/testing
    
    $ cat ~/.odbcinst.ini
    [SQLite Driver]
    Driver          = /opt/homebrew/lib/libsqlite3odbc.dylib
    ```
6. Connections tab > New Connection > SQLite > Test or OK --> **Successful!** 🎉 
    - SQLite Driver shows up now

The above steps didn't work for me on MacOS 13.2 unfortunately. However, installing the intel x86 version of `sqliteodbc` does resolve the issue without making changes to `.Renviron`.

### M1 MacOS 13.2
1. ensure rosetta2 is installed
    - `/usr/sbin/softwareupdate --install-rosetta --agree-to-license`
2. install the x86_64 version of `brew`
    - `arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
3. install the x86_64 version of `sqliteodbc`
    - `arch -x86_64 /usr/local/bin/brew install sqliteodbc`
4. ensure odbc config files are set up correctly 
    ```
    $ cat ~/.odbc.ini
    [SQLite]
    Driver      = SQLite Driver
    Database    = /tmp/testing
    
    $ cat ~/.odbcinst.ini
    [SQLite Driver]
    Driver          = /usr/local/lib/libsqlite3odbc.dylib
    ```
7. Connections tab > New Connection > SQLite > Test or OK --> **Successful!** 🎉

**Comment by sharon-wang** [comment]
I just noticed a discrepancy in my `odbc` library installations in RStudio.

- Packages tab > odbc > DESCRIPTION file > scroll to bottom
    - MacOS 12.6: `Built: R 4.2.0; aarch64-apple-darwin20; 2023-01-17 12:42:20 UTC; unix`
        - note `aarch64`
    - MacOS 13.2: `Built: R 4.2.0; x86_64-apple-darwin17.0; 2023-01-18 13:45:34 UTC; unix`
        - note `x86_64`

I'm looking into why the `x86_64` version is getting installed instead of `aarch64`.

**Comment by rantoun** [comment]
@sharon-wang I had also uninstalled `odbc` and made sure to install using `devtools::install_github("r-dbi/odbc")`. Might be worth trying. I was also on a MacOS 13.2

**Comment by sharon-wang** [comment]
Thanks again @rantoun. I tried your suggestion and still got an x86_64 version, but realized that I had built an x86_64 version of RStudio. Naturally, x86_64 RStudio would install and expect x86_64 libraries and dependencies 😆 

I rebuilt an Arm64 version, installed `odbc` as usual (not the dev build) and was able to reproduce the `Error: nanodbc/nanodbc.cpp:1118: 00000: [unixODBC][Driver Manager]Data source name not found and no default driver specified` error on MacOS 13.2. Adding `ODBCSYSINI=/opt/homebrew/etc` to `.Renviron` resolved the issue.

It seems to me that we should add to the [Drivers documentation](https://solutions.posit.co/connections/db/best-practices/drivers) to set `ODBCSYSINI=/opt/homebrew/etc` in `.Renviron` when running on an Arm-based Mac. Similar documentation could be helpful in https://github.com/r-dbi/odbc as well.
